controller_manager/src/resource_manager.hpp

```diff
@@ -0,0 +1,55 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RESOURCE_MANAGER_HPP_
+#define RESOURCE_MANAGER_HPP_
+
+#include <memory>
+#include <string>
+
+namespace controller_manager
+{
+
+class ResourceStorage;
+
+class ResourceManager
+{
+public:
+  ResourceManager();
+
+  explicit ResourceManager(const std::string & urdf);
+
+  ResourceManager(const ResourceManager &) = delete;
+
+  virtual ~ResourceManager();
+
+  size_t joint_components_size() const;
+
+  size_t sensor_components_size() const;
+
+  size_t actuator_interfaces_size() const;
+
+  size_t sensor_interfaces_size() const;
+
+  size_t system_interfaces_size() const;
+
+  // loan_joint(const std::string & name);
+  // loan_sensor(const std::string & name);
+
+private:
+  std::unique_ptr<ResourceStorage> resource_storage_;
```

Comment:
What is the reason for making this a `unique_ptr`?

Reply:
http://oliora.github.io/2015/12/29/pimpl-and-rule-of-zero.html#scott-meyers-c11-pimpl

Reply:
Do we need this though? Are we going to support ResourceManager inheritance or common interaction at the C++ API level? Is there an ABI to keep stable here? Not really a strong point from me, mostly curious about the motivation.

I'd imagine ResourceManager to be a private as you also [pointed it out](https://github.com/ros-controls/ros2_control/pull/165/files#r496951746) so PIMPL would be better applied to wrap resource manager instead?